pennylane/ops/qubit/matrix_ops.py

```diff
@@ -145,7 +145,8 @@ def __init__(
             )
 
         # If the matrix is sparse, we need to convert it to a csr_matrix
-        if sp.sparse.issparse(U):
+        self._issparse = sp.sparse.issparse(U)
+        if self._issparse:
```

Comment:
@albi3ro, @JerryChen97:
For clarity, I understand sparse/dense decisions are currently at the operator level. Shouldn't this ideally be at the device level?

Reply:
@AmintorDusko `QubitUnitary` updates its own `has_matrix` and `has_sparse_matrix` by checking the input `U`, so `_issparse` used here and shared inside `QubitUnitary`. The `apply_operation` of `default.qubit` decides when to use the sparse pipeline.